### DISCUSSION TITLE
How Can I Keep Native State?

### DISCUSSION THREAD

**Discussion by coder0xff** [fzyzcjy/flutter_rust_bridge#962] [Q&A] [answered]
From what I can gather, references and pointers can't be used as arguments or return values. If, for example, I were to open a database connection in Rust, and the handle struct contained private, non-serializable fields, it becomes pretty difficult to pass it around between Flutter -> Rust invocations. The only workaround I'm aware of is converting pointers to `usize`. And since generics aren't supported, type safety can't be preserved by something like `struct UintPtr<DBHandle>`. Needless to say, that's really undesirable.

Is there a clean way to keep the native state connected to an object that passes through the bridge?

**Comment by fzyzcjy** [comment, marked as answer]
Maybe `opaque`, or `lazy_static`, or `pool`

**Reply by coder0xff** [reply]
I can guess how `pool` would work, using an integer to reference an allocated slot in a pool. I have the basic idea behind `lazy_static` (I'm fairly new to Rust and Flutter/Dart), but I don't know how I would use it here. If my intuition is right, `opaque` seems like it could be what I'm looking for. Is there an example showing it in this use case?

**Reply by fzyzcjy** [reply]
please have a look at the doc :)

**Reply by coder0xff** [reply]
Found it in https://cjycode.com/flutter_rust_bridge/feature/lang_rust_opaque.html
